### ISSUE TITLE
[help] Is there a simple way to convert a sofa file whose GLOBAL_DataType is TF to  FIR

### ISSUE THREAD

**Issue by 907711588** [sofacoustics/API_MO#24]
I see that the programs in the example are all sofa files based on FIR data, but the data format I generated with mesh2hrtf is TF. In SOFA specs 0.6 I saw that TF format is still supported, but I want to know if there is an easy way to convert between TF format and FIR format.

**Comment by 907711588** [comment]
The SOFAconvertConventions.m file in the API_MO folder can do this conversion, but the result of the conversion cannot be saved with SOFAsave.

**Comment by HaHeho** [comment]
In theory it should be straight forward to convert the data and therefore also the convention.

Nevertheless, it of course would be desirable for the provided functions to be directly compatible. What doesn't work for you when transforming or saving?

**Comment by 907711588** [comment]
> In theory it should be straight forward to convert the data and therefore also the convention.
> 
> Nevertheless, it of course would be desirable for the provided functions to be directly compatible. What doesn't work for you when transforming or saving?

Thank you for your reply! My transforming process is as follows:
>> SOFAstart;
>>hrtf_TF = SOFAload( 'E:\simulation\EvaluationGrid.sofa ');
>> hrtf_HRIR = SOFAconvertConventions(hrtf_TF);
>> SOFAsave('E;\hrir.sofa',hrtf_HRIR)

The error is
**Function or variable'ncid' is not recognized**
in NETCDFsave (167 lines)
netcdf.close(ncid);

**Comment by HaHeho** [comment]
> > > SOFAstart;
> > > hrtf_TF = SOFAload( 'E:\simulation\EvaluationGrid.sofa ');
> > > hrtf_HRIR = SOFAconvertConventions(hrtf_TF);
> > > SOFAsave('E;\hrir.sofa',hrtf_HRIR)
> 
> The error is
> **Function or variable'ncid' is not recognized**
> in NETCDFsave (167 lines)
> netcdf.close(ncid);

I'm pretty sure it's supposed to be `SOFAsave('E:\hrir.sofa',hrtf_HRIR);` (colon instead of semicolon).

Don't know if that fixes your problem since it is difficult to verify without an actual data set. Can you provide yours or link to a similar one then?!

**Comment by 907711588** [comment]
> > > > SOFAstart;
> > > > hrtf_TF = SOFAload( 'E:\simulation\EvaluationGrid.sofa ');
> > > > hrtf_HRIR = SOFAconvertConventions(hrtf_TF);
> > > > SOFAsave('E;\hrir.sofa',hrtf_HRIR)
> > 
> > 
> > The error is
> > **Function or variable'ncid' is not recognized**
> > in NETCDFsave (167 lines)
> > netcdf.close(ncid);
> 
> I'm pretty sure it's supposed to be `SOFAsave('E:\hrir.sofa',hrtf_HRIR);` (colon instead of semicolon).
> 
> Don't know if that fixes your problem since it is difficult to verify without an actual data set. Can you provide yours or link to a similar one then?!

emmm，I accidentally made a mistake just now. The real error is Error processing N.
Here is  my data produced by mesh2HRTF:
[EvaluationGrid.zip](https://github.com/sofacoustics/API_MO/files/5745729/EvaluationGrid.zip)

**Comment by HaHeho** [comment]
Yeah, that is what I encountered when testing another data set as well. This script helps to debug shows where the error actually happens and helps to fix it.

```Matlab
close all; clear; clc;
SOFAstart;

%%
% load reference HRIR
ref_HRIR = SOFAload(fullfile(SOFAdbPath,'database','ari','hrtf_NH4.sofa'));

% print HRIR API Dimensions
%                  ListenerPosition: 'IC'
%                  ReceiverPosition: 'RCI'
%                    SourcePosition: 'MC'
%                   EmitterPosition: 'ECI'
%                        ListenerUp: 'IC'
%                      ListenerView: 'IC'
%     MeasurementSourceAudioChannel: 'M'
%           MeasurementAudioLatency: 'MR'
%                              Data: [1×1 struct]
ref_HRIR.API.Dimensions

% test saving reference HRIR ... works
SOFAsave('ref_HRIR.sofa',ref_HRIR)

%%
% load desired HRTF
hrtf_TF = SOFAload('EvaluationGrid.sofa');

% transform into HRIR
hrtf_HRIR = SOFAconvertConventions(hrtf_TF);
% print HRIR API Dimensions
%     ListenerPosition: 'IC'
%     ReceiverPosition: 'RCI'
%       SourcePosition: 'MC'
%      EmitterPosition: 'ECI'
%           ListenerUp: 'IC'
%         ListenerView: 'IC'
%                    N: 'N'
%                 Data: [1×1 struct]
hrtf_HRIR.API.Dimensions

% % save desired HRIR ... doesn't work due to
% % Error using NETCDFsave (line 172)
% % Error processing N
% % Error message: Reference to non-existent field 'N'.
% SOFAsave('EvaluationGrid_HRIR.sofa',hrtf_HRIR)

% remove disruptive API dimension
% (I don't know what is supposed to be there according to the convention)
hrtf_HRIR.API.Dimensions = rmfield(hrtf_HRIR.API.Dimensions,'N');
hrtf_HRIR.API.Dimensions

% save desired HRIR ... should also work now
SOFAsave('EvaluationGrid_HRIR.sofa',hrtf_HRIR)
```

@petibub Is this due to either data set not being according to convention? Or is this a bug in `SOFAconvertConventions()` or `SOFAsave()`?

**Comment by HaHeho** [comment]
For reference, I used the API `v 1.1.1` (from https://sourceforge.net/projects/sofacoustics/files/?source=navbar).
I believe the last code state from GitHub won't be compatible since it is being prepared for the new SOFA 2.0 release.

@petibub It is confusing that the most recent releases of the Matlab API are only available on Sourceforge and not on GitHub. It would be good if the API releases are available in one place exclusively. IMHO this should be as release tags in the repository where the actual development code is hosted.
But there are probably plenty of resources linking to either GitHub or Sourceforge by now. So withdrawing from either reference might not be convenient for users either. Nevertheless, I believe the newer releases should be made available on GitHub as well?

**Comment by 907711588** [comment]
> Yeah, that is what I encountered when testing another data set as well. This script helps to debug shows where the error actually happens and helps to fix it.
> 
> ```matlab
> close all; clear; clc;
> SOFAstart;
> 
> %%
> % load reference HRIR
> ref_HRIR = SOFAload(fullfile(SOFAdbPath,'database','ari','hrtf_NH4.sofa'));
> 
> % print HRIR API Dimensions
> %                  ListenerPosition: 'IC'
> %                  ReceiverPosition: 'RCI'
> %                    SourcePosition: 'MC'
> %                   EmitterPosition: 'ECI'
> %                        ListenerUp: 'IC'
> %                      ListenerView: 'IC'
> %     MeasurementSourceAudioChannel: 'M'
> %           MeasurementAudioLatency: 'MR'
> %                              Data: [1×1 struct]
> ref_HRIR.API.Dimensions
> 
> % test saving reference HRIR ... works
> SOFAsave('ref_HRIR.sofa',ref_HRIR)
> 
> %%
> % load desired HRTF
> hrtf_TF = SOFAload('EvaluationGrid.sofa');
> 
> % transform into HRIR
> hrtf_HRIR = SOFAconvertConventions(hrtf_TF);
> % print HRIR API Dimensions
> %     ListenerPosition: 'IC'
> %     ReceiverPosition: 'RCI'
> %       SourcePosition: 'MC'
> %      EmitterPosition: 'ECI'
> %           ListenerUp: 'IC'
> %         ListenerView: 'IC'
> %                    N: 'N'
> %                 Data: [1×1 struct]
> hrtf_HRIR.API.Dimensions
> 
> % % save desired HRIR ... doesn't work due to
> % % Error using NETCDFsave (line 172)
> % % Error processing N
> % % Error message: Reference to non-existent field 'N'.
> % SOFAsave('EvaluationGrid_HRIR.sofa',hrtf_HRIR)
> 
> % remove disruptive API dimension
> % (I don't know what is supposed to be there according to the convention)
> hrtf_HRIR.API.Dimensions = rmfield(hrtf_HRIR.API.Dimensions,'N');
> hrtf_HRIR.API.Dimensions
> 
> % save desired HRIR ... should also work now
> SOFAsave('EvaluationGrid_HRIR.sofa',hrtf_HRIR)
> ```
> 
> @petibub Is this due to either data set not being according to convention? Or is this a bug in `SOFAconvertConventions()` or `SOFAsave()`?

Thanks for your help！I can save it normally through your script.

**Comment by petibub** [comment]
> I see that the programs in the example are all sofa files based on FIR data, but the data format I generated with mesh2hrtf is TF. In SOFA specs 0.6 I saw that TF format is still supported, but I want to know if there is an easy way to convert between TF format and FIR format.

Dear 907711588, 

The function SOFAconvertConventions.m from the API version 1.1.1 does **not** support conversions **to** SimpleFreeFieldTF. When you look into  SOFAconvertConventions.m, line 2, it says: 

`% SOFAconvertConventions(OBJ) converts an object to SimpleFreeFieldHRIR`

Sorry to say that. 

But, if you want to convert to TF, you could look at [demo_FreeFieldHRTF.m from the commit 0e58210](https://github.com/sofacoustics/API_MO/blob/0e58210bf1811a104514bf4e26fc3fd4e0519466/API_MO/demos/demo_FreeFieldHRTF.m). Between lines 24 and 56, I convert a SingleFreeFieldHRIR to SingleFreeFieldTF data structure and save the new structure as a SOFA file. Hope that helps.

Note that the github files are thought to be a working repository - nothing is supposed to be tested or even be considered as a release. Especially now, we're preparing for the update for SOFA 2.0, so many things can be broken. For a working version of the API, please refer to the download link at [sourceforge](https://sourceforge.net/projects/sofacoustics/files/). 

Piotr

**Comment by petibub** [comment]
> For reference, I used the API `v 1.1.1` (from https://sourceforge.net/projects/sofacoustics/files/?source=navbar).
> I believe the last code state from GitHub won't be compatible since it is being prepared for the new SOFA 2.0 release.^

Dear HaHeho, 

that's exactly what's happening currently - we're preparing the API for SOFA 2.0, so the working repository here is not thought to be used for work. 


> @petibub It is confusing that the most recent releases of the Matlab API are only available on Sourceforge and not on GitHub. It would be good if the API releases are available in one place exclusively. 

You're right, I forgot that we have some "zombie" releases here on github. Thanks for the pointer - those are gone now. 

The system works as follows: 
- development on github
- releases on sourceforge

> IMHO this should be as release tags in the repository where the actual development code is hosted.
> But there are probably plenty of resources linking to either GitHub or Sourceforge by now. So withdrawing from either reference might not be convenient for users either. Nevertheless, I believe the newer releases should be made available on GitHub as well?

Unfortunately, years ago, I had many issues with people not being able to find the "download" button at github. I think that we all agree, that the "download a release" is not very prominent at github - github is more programmer-oriented. That's why I rather like to keep the releases at Sourceforge - where people can find a big shiny "Download" button. But I love github as a code repository, so here we are :-)

Best regards, 
Piotr

**Comment by petibub** [comment]
> emmm，I accidentally made a mistake just now. The real error is Error processing N.
> Here is my data produced by mesh2HRTF:
> [EvaluationGrid.zip](https://github.com/sofacoustics/API_MO/files/5745729/EvaluationGrid.zip)

Dear 907711588, 

your data have some problems; They contain an attribute with the name `GLOBAL__NCProperties`. The double-underscore is forbidden in SOFA, thus an error on save. Try: `X2=SOFAload('EvaluationGrid.sofa'); SOFAsave('test.sofa',X2);` and you'll the error without any conversions. Remove this illegal attribute, i.e., `X2=rmfield(X2,'GLOBAL__NCProperties');` to be able to save to a SOFA file. Interestingly, our SOFA files resulting from Mesh2HRTF, at least those created on 30.6.2020. Mesh2HRTF had some major updates since then - maybe that's a problem. 

Also, there is a bug in SOFAconvertConventions: in that file, I forgot to remove the `N` from the `API.Dimensions` structure. 

In summary, try the following: 
`X=SOFAload('EvaluationGrid.sofa');`
`X=rmfield(X,'GLOBAL__NCProperties'); % fix the problem of Mesh2HRTF`
`Y=SOFAconvertConventions(X); `
`Y.API.Dimensions=rmfield(Y.API.Dimensions,'N'); % fix the problem of SOFAconvertConventions 1.1.1`
`Y=SOFAupdateDimensions(Y);`
`SOFAsave('test.sofa',Y);`

This saves your data as SimpleFreeFieldHRIR my machine (Windows 10, Matlab R2016a) without errors.

Hope that helps, 
Piotr
